Fix multi-line strings bug

joshuakb2 opened the pull request:
Fix bug #16 

This:
```
main =
    <div>Hi</div>
```
should become
```
main =
    Html.node "div" [] [Html.text "Hi"]
```
and this:
```
main =
    <p>
        foo
    </p>
```
should become
```
main =
    Html.node "p" [] [Html.text """
        foo
    """]
```

joshuakb2 committed:
Fix for multiline string bug.

pzavolinsky commented on src/generator.ts:
```diff
@@ -47,7 +47,11 @@ function generateExpression(expr:Expression) {
     case 'code': return expr.value;
-    case 'text': return `Html.text "${expr.value.replace(/"/g, '\\"')}"`;
+    case 'text': {
+      let multiline = expr.value.match(/\n|\r/) ? true : false;
```
```ts
const quote = expr.value.includes(/\n|\r/) ? '"""' : '"';
```

pzavolinsky commented on src/generator.ts:
```diff
@@ -47,7 +47,11 @@ function generateExpression(expr:Expression) {
+    case 'text': {
+      let multiline = expr.value.match(/\n|\r/) ? true : false;
+      let quote = multiline ? '"""' : '"';
+      return `Html.text ${quote}${expr.value.replace(/"/g, '\\"')}${quote}`;
```
Also, here....
```ts
const escapedQuote = quote.replace(/"/g, '\\"');
return Html.text ${quote}${expr.value.replace(new RegExp(quote, 'g'), escapedQuote)}${quote};
```

:point_up: see if this fixes the failing test (and allows you to drop the `\`s in the cucumber steps (since the actual output we should produce should be `Html.text """` and not `Html.text \"\"\"`).

pzavolinsky commented:
Hey @joshuakb2 thanks a lot for this contrib!

I left you two comments in the PR above, one of which I believe should fix the broken test.

Could you also add more tests for mixed content (e.g. a multiline node with other multiline inside)?

Let me know when I can review again (also please squash all commits down to a single one).

Thanks again,
Cheers!

joshuakb2 commented:
I'd be happy to make those changes, but I'm not sure which test is failing.

Here's what I'm doing to build -- have I missed something?
```
joshua@Joshua-X1:~/projects/elmx$ npm ci
added 495 packages in 3.776s
joshua@Joshua-X1:~/projects/elmx$ ./node_modules/.bin/gulp 
[14:08:15] Using gulpfile ~/projects/elmx/gulpfile.js
[14:08:15] Starting 'clean'...
[14:08:15] Starting 'copy-bundle'...
[14:08:15] Finished 'copy-bundle' after 4.1 ms
[14:08:15] Finished 'clean' after 20 ms
[14:08:15] Starting 'build'...
[14:08:16] Finished 'build' after 1.21 s
[14:08:16] Starting 'test'...
......................................................................................................

49 scenarios (49 passed)
102 steps (102 passed)
0m00.011s
[14:08:17] Finished 'test' after 451 ms
[14:08:17] Starting 'bundle'...
[14:08:17] Finished 'bundle' after 392 ms
[14:08:17] Starting 'uglify'...
[14:08:17] Finished 'uglify' after 843 μs
[14:08:17] Starting 'default'...
[14:08:17] Finished 'default' after 13 μs
joshua@Joshua-X1:~/projects/elmx$
```

joshuakb2 commented:
Oh, I'm sorry, you're referring to the one that I `@ignore`d. The issue with that one actually has nothing to do with the triple quotes. The generator doesn't seem to output any whitespace output in that case, before, after, or between the two text variables.

joshuakb2 commented:
I made the first change you suggested, and I added a comment to full.feature to clarify some confusing syntax. I also squashed everything into one commit.

The test scenario I added with an `@ignore` tag doesn't have anything to do with the multi-line bug -- I can reproduce the problem without multi-line input. However, I took a look at how Babel handles whitespace in JSX, and it looks like any time there is a new line in a sequence of whitespace characters, all of the whitespace is omitted (or replaced with a single space if the whitespace occurs between two bits of visible text).

This difference in behavior leads me to think that we shouldn't be outputting multi-line Elm syntax at all. I think implementing this differently might also lead to a better solution to the failing test case I added.

Feel free to review what I've done so far, but I'm going to continue working on this in the meantime.

pzavolinsky commented:
Hey @joshuakb2 thanks a lot for your time and your work.

Let me know how the research goes.

(It was a great idea, IMO, to see how other similar conversions work (e.g. HTML, JSX, etc.))

Cheers!
